# Importing Libraries

In [17]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [18]:
# Importing “opm_redesigned.pkl” data set from the “Prepared Data” folder as ords_prods_merge by creating a string of the path

path = r'C:\Users\rbhte\OneDrive\Documents\Achievement_4_Monami\Instacart Basket Analysis'
print (path)
print ( '..................................................................................')
# Correctly reading the pickle file
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'opm_redesigned.pkl'))

print(ords_prods_merge.columns)

C:\Users\rbhte\OneDrive\Documents\Achievement_4_Monami\Instacart Basket Analysis
..................................................................................
Index(['order_id', 'user_id', 'order_number', 'order_day_of_week',
       'order_hour_of_day', 'days_since_prior_order', 'new_customer',
       'product_id', 'add_to_cart_order', 'reordered', 'product_name',
       'aisle_id', 'department_id', 'prices', 'price_label', 'busiest_day',
       'day_category', 'busiest_period_of_day'],
      dtype='object')


In [19]:
# Create a subset with the first one million entries
df = ords_prods_merge.head(1000000)

# Print confirmation and the shape of the new subset
print(f"Subset 'df' created with {df.shape[0]} rows and {df.shape[1]} columns.")

Subset 'df' created with 1000000 rows and 18 columns.


In [20]:
df.head(10)

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2539329        1             1                  2                  8   
2   2539329        1             1                  2                  8   
3   2539329        1             1                  2                  8   
4   2539329        1             1                  2                  8   
5   2398795        1             2                  3                  7   
6   2398795        1             2                  3                  7   
7   2398795        1             2                  3                  7   
8   2398795        1             2                  3                  7   
9   2398795        1             2                  3                  7   

   days_since_prior_order  new_customer  product_id  add_to_cart_order  \
0                     NaN          True         196                  1   
1                     NaN          True       14084                  2   
2                     NaN          True       12427                  3   
3                     NaN          True       26088                  4   
4                     NaN          True       26405                  5   
5                    15.0         False         196                  1   
6                    15.0         False       10258                  2   
7                    15.0         False       12427                  3   
8                    15.0         False       13176                  4   
9                    15.0         False       26088                  5   

   reordered                             product_name  aisle_id  \
0          0                                     Soda        77   
1          0  Organic Unsweetened Vanilla Almond Milk        91   
2          0                      Original Beef Jerky        23   
3          0               Aged White Cheddar Popcorn        23   
4          0         XL Pick-A-Size Paper Towel Rolls        54   
5          1                                     Soda        77   
6          0                               Pistachios       117   
7          1                      Original Beef Jerky        23   
8          0                   Bag of Organic Bananas        24   
9          1               Aged White Cheddar Popcorn        23   

   department_id  prices        price_label     busiest_day  day_category  \
0              7     9.0  Mid-range product  Regularly busy   normal_days   
1             16    12.5  Mid-range product  Regularly busy   normal_days   
2             19     4.4  Low-range product  Regularly busy   normal_days   
3             19     4.7  Low-range product  Regularly busy   normal_days   
4             17     1.0  Low-range product  Regularly busy   normal_days   
5              7     9.0  Mid-range product  Regularly busy  slowest_days   
6             19     3.0  Low-range product  Regularly busy  slowest_days   
7             19     4.4  Low-range product  Regularly busy  slowest_days   
8              4    10.3  Mid-range product  Regularly busy  slowest_days   
9             19     4.7  Low-range product  Regularly busy  slowest_days   

  busiest_period_of_day  
0        Average orders  
1        Average orders  
2        Average orders  
3        Average orders  
4        Average orders  
5        Average orders  
6        Average orders  
7        Average orders  
8        Average orders  
9        Average orders

In [5]:
# Group data by 'department_id' and calculate the mean of 'order_number' for each group using the agg() function
df.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                14.800024
2                17.091743
3                17.913544
4                17.893092
5                15.214270
6                15.382135
7                17.694027
8                16.458105
9                15.957363
10               20.091818
11               16.482026
12               15.615061
13               16.484023
14               17.524632
15               15.691875
16               18.014071
17               16.150593
18               19.602850
19               17.631340
20               17.138607
21               21.956893

In [21]:
# Using the mean() function to achieve similar results
df.groupby('department_id')['order_number'].mean()

department_id
1     14.800024
2     17.091743
3     17.913544
4     17.893092
5     15.214270
6     15.382135
7     17.694027
8     16.458105
9     15.957363
10    20.091818
11    16.482026
12    15.615061
13    16.484023
14    17.524632
15    15.691875
16    18.014071
17    16.150593
18    19.602850
19    17.631340
20    17.138607
21    21.956893
Name: order_number, dtype: float64

In [22]:
# Produce multiple statistics at the same time
df.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})

order_number        
                      mean min max
department_id                     
1                14.800024   1  99
2                17.091743   1  98
3                17.913544   1  99
4                17.893092   1  99
5                15.214270   1  99
6                15.382135   1  99
7                17.694027   1  99
8                16.458105   1  91
9                15.957363   1  99
10               20.091818   1  99
11               16.482026   1  99
12               15.615061   1  99
13               16.484023   1  99
14               17.524632   1  99
15               15.691875   1  99
16               18.014071   1  99
17               16.150593   1  99
18               19.602850   1  99
19               17.631340   1  99
20               17.138607   1  99
21               21.956893   1  97

In [23]:
# Create a copy of the ords_prods_merge DataFrame
ords_prods_merge_copy = ords_prods_merge.copy()

# Confirm the copy was created successfully
print("DataFrame copy created successfully.")

DataFrame copy created successfully.


In [24]:
# Produce multiple statistics (mean, min, max) at the same time
ords_prods_merge_copy.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})

order_number        
                      mean min max
department_id                     
1                15.457838   1  99
2                17.277920   1  99
3                17.170395   1  99
4                17.811403   1  99
5                15.215751   1  99
6                16.439806   1  99
7                17.225802   1  99
8                15.340650   1  99
9                15.895474   1  99
10               20.197148   1  99
11               16.170638   1  99
12               15.887671   1  99
13               16.583536   1  99
14               16.773669   1  99
15               16.165037   1  99
16               17.665606   1  99
17               15.694469   1  99
18               19.310397   1  99
19               17.177343   1  99
20               16.473447   1  99
21               22.902379   1  99

 Analysis of Results: Full Dataset vs Subset

 Key Observations:
 
1. **Mean Values:**
   - The mean `order_number` values for most `department_id`s in the full dataset (`ords_prods_merge_copy`) are **slightly higher** compared to those in the subset (`df`).
   - For example:
     - `department_id 1`: Mean increases from **14.80** (subset) to **15.46** (full dataset).
     - `department_id 6`: Mean increases from **15.38** (subset) to **16.44** (full dataset).
     - This suggests that the subset may underrepresent users with higher order numbers, leading to slightly lower averages.

2. **Minimum and Maximum Values:**
   - The minimum `order_number` value for all departments is **1** in both datasets, indicating consistency.
   - The maximum `order_number` value is **99** for nearly all departments in both datasets, except for `department_id 21`, where:
     - Subset maximum = **97**
     - Full dataset maximum = **99**

3. **Variability Across Departments:**
   - Departments with higher mean values, such as `department_id 10` and `department_id 21`, show slight increases in the full dataset compared to the subset.
   - Departments with lower mean values, such as `department_id 1` and `department_id 5`, also exhibit increases but to a smaller extent.

4. **Subset Representativeness:**
   - The subset (`df`) appears to be a reasonable representation of the overall dataset in terms of distribution. However, the subset may slightly **underrepresent users with higher order numbers** in some departments, leading to marginally lower means.

  Conclusion:
  
- The analysis shows that while the subset (`df`) provides a similar overall pattern, it **slightly underestimates mean order numbers** for certain departments compared to the full dataset (`ords_prods_merge_copy`).
- For the most accurate insights, it is recommended to use the **full dataset** to ensure all patterns and variations are captured.


In [25]:
### Create a flag using the following criteria
# If the maximum orders the user has made is over 40, then the customer will be labeled a “Loyal customer.”
# If the maximum orders the user has made is over 10 but less than or equal to 40, then the customer will be labeled a “Regular customer.”
# If the maximum orders the user has made is less than or equal to 10, then the customer will be labeled a “New customer.”

# Create a new column 'max_order' with the maximum 'order_number' for each 'user_id'
ords_prods_merge_copy['max_order'] = ords_prods_merge_copy.groupby(['user_id'])['order_number'].transform('max')

# Confirm the column is created successfully
print("Column 'max_order' created successfully.")

Column 'max_order' created successfully.


In [26]:
ords_prods_merge_copy['max_order'].value_counts(dropna = False)

max_order
99    1171333
8      811843
6      811396
9      810213
7      803979
       ...   
97      44949
98      44587
96      40453
2           6
1           5
Name: count, Length: 99, dtype: int64

In [27]:
ords_prods_merge_copy.head(9)

order_id  user_id  order_number  order_day_of_week  order_hour_of_day  \
0   2539329        1             1                  2                  8   
1   2539329        1             1                  2                  8   
2   2539329        1             1                  2                  8   
3   2539329        1             1                  2                  8   
4   2539329        1             1                  2                  8   
5   2398795        1             2                  3                  7   
6   2398795        1             2                  3                  7   
7   2398795        1             2                  3                  7   
8   2398795        1             2                  3                  7   

   days_since_prior_order  new_customer  product_id  add_to_cart_order  \
0                     NaN          True         196                  1   
1                     NaN          True       14084                  2   
2                     NaN          True       12427                  3   
3                     NaN          True       26088                  4   
4                     NaN          True       26405                  5   
5                    15.0         False         196                  1   
6                    15.0         False       10258                  2   
7                    15.0         False       12427                  3   
8                    15.0         False       13176                  4   

   reordered                             product_name  aisle_id  \
0          0                                     Soda        77   
1          0  Organic Unsweetened Vanilla Almond Milk        91   
2          0                      Original Beef Jerky        23   
3          0               Aged White Cheddar Popcorn        23   
4          0         XL Pick-A-Size Paper Towel Rolls        54   
5          1                                     Soda        77   
6          0                               Pistachios       117   
7          1                      Original Beef Jerky        23   
8          0                   Bag of Organic Bananas        24   

   department_id  prices        price_label     busiest_day  day_category  \
0              7     9.0  Mid-range product  Regularly busy   normal_days   
1             16    12.5  Mid-range product  Regularly busy   normal_days   
2             19     4.4  Low-range product  Regularly busy   normal_days   
3             19     4.7  Low-range product  Regularly busy   normal_days   
4             17     1.0  Low-range product  Regularly busy   normal_days   
5              7     9.0  Mid-range product  Regularly busy  slowest_days   
6             19     3.0  Low-range product  Regularly busy  slowest_days   
7             19     4.4  Low-range product  Regularly busy  slowest_days   
8              4    10.3  Mid-range product  Regularly busy  slowest_days   

  busiest_period_of_day  max_order  
0        Average orders         10  
1        Average orders         10  
2        Average orders         10  
3        Average orders         10  
4        Average orders         10  
5        Average orders         10  
6        Average orders         10  
7        Average orders         10  
8        Average orders         10

In [28]:
# Create a flag that assigns a “loyalty” label to a user ID based on its corresponding max order value
ords_prods_merge_copy.loc[ords_prods_merge_copy['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'
ords_prods_merge_copy.loc[(ords_prods_merge_copy['max_order'] <= 40) & (ords_prods_merge_copy['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'
ords_prods_merge_copy.loc[ords_prods_merge_copy['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [29]:
# Print the frequency of the new “loyalty_flag” column
ords_prods_merge_copy.groupby(['loyalty_flag']).user_id.nunique()

loyalty_flag
Loyal customer       17017
New customer        112328
Regular customer     76864
Name: user_id, dtype: int64

In [30]:
ords_prods_merge_copy[['user_id', 'loyalty_flag', 'order_number']].head(6)

user_id  loyalty_flag  order_number
0        1  New customer             1
1        1  New customer             1
2        1  New customer             1
3        1  New customer             1
4        1  New customer             1
5        1  New customer             2

In [31]:
# Group the data by 'loyalty_flag' and calculate basic statistics for the 'prices' column
loyalty_price_stats = ords_prods_merge_copy.groupby('loyalty_flag')['prices'].describe()

# Calculate the unique user count for each loyalty flag
unique_user_count = ords_prods_merge_copy.groupby('loyalty_flag')['user_id'].nunique()

# Combine the statistics with the unique user count
loyalty_price_stats['unique_user_count'] = unique_user_count

# Display the updated statistics
print(loyalty_price_stats)

                       count       mean         std  min  25%  50%   75%  \
loyalty_flag                                                               
Loyal customer    10284093.0  10.386333  327.981293  1.0  4.2  7.4  11.2   
New customer       6243990.0  13.294672  597.521545  1.0  4.2  7.4  11.3   
Regular customer  15876776.0  12.495719  539.679504  1.0  4.2  7.4  11.3   

                      max  unique_user_count  
loyalty_flag                                  
Loyal customer    99999.0              17017  
New customer      99999.0             112328  
Regular customer  99999.0              76864  


In [32]:
# Calculate the mean price for each user and assign a spending flag
ords_prods_merge_copy['spending_flag'] = ords_prods_merge_copy.groupby('user_id')['prices'].transform('mean').apply(
    lambda x: 'Low spender' if x < 10 else 'High spender'
)

# Calculate the unique user count for each spending flag
unique_user_count = ords_prods_merge_copy.groupby('spending_flag')['user_id'].nunique()

# Display the unique user count
print("Unique User Count by Spending Flag:")
print(unique_user_count)

Unique User Count by Spending Flag:
spending_flag
High spender      5357
Low spender     200852
Name: user_id, dtype: int64


In [33]:
# Create the 'order_frequency_flag' column
ords_prods_merge_copy['order_frequency_flag'] = ords_prods_merge_copy.groupby('user_id')['days_since_prior_order'].transform('median').apply(
    lambda x: 'Non-frequent customer' if x > 20 else 
              'Regular customer' if 10 < x <= 20 else 
              'Frequent customer'
)

# Calculate unique user IDs in each group
unique_user_count = ords_prods_merge_copy.groupby('order_frequency_flag')['user_id'].nunique()

# Display the unique user count
print("Unique User Count by Order Frequency Flag:")
print(unique_user_count)

Unique User Count by Order Frequency Flag:
order_frequency_flag
Frequent customer        86597
Non-frequent customer    59619
Regular customer         59993
Name: user_id, dtype: int64


In [34]:
ords_prods_merge_copy.columns

Index(['order_id', 'user_id', 'order_number', 'order_day_of_week',
       'order_hour_of_day', 'days_since_prior_order', 'new_customer',
       'product_id', 'add_to_cart_order', 'reordered', 'product_name',
       'aisle_id', 'department_id', 'prices', 'price_label', 'busiest_day',
       'day_category', 'busiest_period_of_day', 'max_order', 'loyalty_flag',
       'spending_flag', 'order_frequency_flag'],
      dtype='object')

In [35]:
ords_prods_merge_copy['prices'].max()

99999.0

In [36]:
ords_prods_merge_copy['prices'].mean()

11.980226

In [37]:
ords_prods_merge_copy['prices'].median()

7.400000095367432

In [38]:
ords_prods_merge_copy.loc[ords_prods_merge_copy['prices'] > 100]

order_id  user_id  order_number  order_day_of_week  \
1576        912404       17            12                  2   
1638        603376       17            22                  6   
16522      3264360      135             2                  2   
16528       892534      135             3                  0   
53672       229704      342             8                  1   
...            ...      ...           ...                ...   
32320511   3172853   205650            18                  1   
32347717   2504315   205818             3                  5   
32347727   1108388   205818             5                  4   
32380527   1916142   206049             1                  2   
32380551    379732   206049             4                  1   

          order_hour_of_day  days_since_prior_order  new_customer  product_id  \
1576                     14                     5.0         False       21553   
1638                     16                     4.0         False       21553   
16522                    21                    13.0         False       21553   
16528                     8                    12.0         False       21553   
53672                    19                    30.0         False       21553   
...                     ...                     ...           ...         ...   
32320511                  9                     7.0         False       21553   
32347717                 15                     3.0         False       21553   
32347727                  5                     1.0         False       21553   
32380527                 17                     NaN          True       21553   
32380551                 14                     5.0         False       21553   

          add_to_cart_order  reordered  ... department_id   prices  \
1576                      5          0  ...            16  14900.0   
1638                      3          1  ...            16  14900.0   
16522                     6          0  ...            16  14900.0   
16528                     3          1  ...            16  14900.0   
53672                     9          0  ...            16  14900.0   
...                     ...        ...  ...           ...      ...   
32320511                 17          1  ...            16  14900.0   
32347717                 13          0  ...            16  14900.0   
32347727                  5          1  ...            16  14900.0   
32380527                  2          0  ...            16  14900.0   
32380551                  4          1  ...            16  14900.0   

                 price_label     busiest_day  day_category  \
1576      High-range product  Regularly busy   normal_days   
1638      High-range product  Regularly busy   normal_days   
16522     High-range product  Regularly busy   normal_days   
16528     High-range product     Busiest day  busiest_days   
53672     High-range product  Regularly busy  busiest_days   
...                      ...             ...           ...   
32320511  High-range product  Regularly busy  busiest_days   
32347717  High-range product  Regularly busy   normal_days   
32347727  High-range product      Least busy  slowest_days   
32380527  High-range product  Regularly busy   normal_days   
32380551  High-range product  Regularly busy  busiest_days   

         busiest_period_of_day max_order      loyalty_flag  spending_flag  \
1576               Most orders        40  Regular customer   High spender   
1638            Average orders        40  Regular customer   High spender   
16522           Average orders         4      New customer   High spender   
16528           Average orders         4      New customer   High spender   
53672           Average orders        16  Regular customer   High spender   
...                        ...       ...               ...            ...   
32320511           Most orders        25  Regular customer   High spender   
32347717           Most orders        25  Regular customer   High 

In [39]:
ords_prods_merge_copy.loc[ords_prods_merge_copy['prices'] >100, 'prices'] = np.nan

In [40]:
ords_prods_merge_copy['prices'].max()

25.0

In [41]:
ords_prods_merge_copy['prices'].value_counts(dropna = False)

prices
12.300000    624261
10.300000    510496
4.000000     447272
6.300000     438467
1.300000     425691
              ...  
20.200001       123
22.700001        99
21.500000        85
18.299999        11
21.000000         8
Name: count, Length: 241, dtype: int64

In [42]:
# Downcasting numerical columns to reduce memory usage in ords_prods_merge_copy
for column in ords_prods_merge_copy.columns:
    if ords_prods_merge_copy[column].dtype == 'float64':
        ords_prods_merge_copy[column] = pd.to_numeric(ords_prods_merge_copy[column], downcast='float')
    elif ords_prods_merge_copy[column].dtype == 'int64':
        ords_prods_merge_copy[column] = pd.to_numeric(ords_prods_merge_copy[column], downcast='integer')

print("Downcasting completed for ords_prods_merge_copy.")

Downcasting completed for ords_prods_merge_copy.


In [43]:
# Check the data types and other info for ords_prods_merge_copy
print("Data types and other information for ords_prods_merge_copy:")
ords_prods_merge_copy.info()

Data types and other information for ords_prods_merge_copy:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32404859 entries, 0 to 32404858
Data columns (total 22 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   order_id                int32  
 1   user_id                 int32  
 2   order_number            int8   
 3   order_day_of_week       int8   
 4   order_hour_of_day       int8   
 5   days_since_prior_order  float32
 6   new_customer            bool   
 7   product_id              int32  
 8   add_to_cart_order       int16  
 9   reordered               int8   
 10  product_name            object 
 11  aisle_id                int16  
 12  department_id           int8   
 13  prices                  float32
 14  price_label             object 
 15  busiest_day             object 
 16  day_category            object 
 17  busiest_period_of_day   object 
 18  max_order               int8   
 19  loyalty_flag            object 
 20  spendi

In [44]:
# Export the merged dataframe in pickle format

ords_prods_merge_copy.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge_4.8improved.pkl'))
print("ords_prods_merge_4.8improved.pkl created successfully.")

ords_prods_merge_4.8improved.pkl created successfully.
